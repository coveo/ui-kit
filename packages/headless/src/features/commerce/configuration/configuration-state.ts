import {CoreConfigurationState} from '../../configuration/configuration-state.js';

export interface ConfigurationState extends CoreConfigurationState {
  /**
   * The global headless engine Commerce API configuration.
   */
  commerce: CommerceState;
}

interface CommerceState {
  /**
   * The Commerce API base URL to use.
   *
   * By default, will append `/rest/organizations/{organizationId}/commerce/v2` to the automatically resolved
   * platform [organization endpoint](https://docs.coveo.com/en/mcc80216)
   * (i.e., `https;://<ORG_ID>.org<hipaa|dev|stg|>.coveo.com`)
   *
   * If necessary, you can override this value by specifying a `commerce.proxyBaseUrl` in the configuration of your
   * commerce engine, or by manually dispatching the `updateProxyBaseUrl` action.
   */
  apiBaseUrl?: string;
}

export const getConfigurationInitialState: () => ConfigurationState = () => ({
  accessToken: '',
  environment: 'prod',
  organizationId: '',
  analytics: {
    enabled: true,
    trackingId: '',
    source: {},
  },
  commerce: {},
});
