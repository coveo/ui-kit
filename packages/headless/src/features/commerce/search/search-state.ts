import {CommerceAPIErrorStatusResponse} from '../../../api/commerce/commerce-api-error-response';
import {Product} from '../../../api/commerce/common/product';
import {AnyFacetResponse} from '../facets/facet-set/interfaces/response';

export interface CommerceSearchState {
  error: CommerceAPIErrorStatusResponse | null;
  isLoading: boolean;
  requestId: string;
  responseId: string;
  products: Product[];
  facets: AnyFacetResponse[];
  queryExecuted: string;
}

export const getCommerceSearchInitialState = (): CommerceSearchState => ({
  error: null,
  isLoading: false,
  requestId: '',
  responseId: '',
  products: [],
  facets: [],
  queryExecuted: '',
});
