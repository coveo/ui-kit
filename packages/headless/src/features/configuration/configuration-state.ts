import {platformUrl} from '../../api/platform-client';
import {IRuntimeEnvironment} from 'coveo.analytics';

export interface ConfigurationState {
  /**
   * The unique identifier of the target Coveo Cloud organization (e.g., `mycoveocloudorganizationg8tp8wu3`)
   */
  organizationId: string;
  /**
   * The access token to use to authenticate requests against the Coveo Cloud endpoints. Typically, this will be an API key or search token that grants the privileges to execute queries and push usage analytics data in the target Coveo Cloud organization.
   */
  accessToken: string;
  /**
   * The Plaform URL to use.
   * By default, https://platform.cloud.coveo.com
   */
  platformUrl: string;
  /**
   * The global headless engine Search API configuration.
   */
  search: {
    /**
     * The Search API base URL to use.
     * By default, will append /rest/search/v2 to the platformUrl value.
     */
    apiBaseUrl: string;
    /**
     * The locale of the current user. Must comply with IETF’s BCP 47 definition: https://www.rfc-editor.org/rfc/bcp/bcp47.txt.
     */
    locale: string;
  };
  /**
   * The global headless engine Usage Analytics API configuration.
   */
  analytics: {
    /**
     * Specifies if analytics tracking should be enabled. By default analytics events are tracked.
     */
    enabled: boolean;
    /**
     * The Analytics API base URL to use.
     * By default, will append /rest/ua to the platformUrl value.
     */
    apiBaseUrl: string;
    /**
     * Origin level 2 is a usage analytics event metadata whose value should typically be the name/identifier of the tab from which the usage analytics event originates.
     *
     * When logging a Search usage analytics event, originLevel2 should always be set to the same value as the corresponding tab (parameter) Search API query parameter so Coveo Machine Learning models function properly, and usage analytics reports and dashboards are coherent.
     *
     * This value is optional, and will automatically try to resolve itself from the tab search parameter.
     */
    originLevel2: string;

    /**
     * Origin level 3 is a usage analytics event metadata whose value should typically be the URL of the page that linked to the search interface that’s making the request.
     *
     * When logging a Search usage analytics event, originLevel3 should always be set to the same value as the corresponding referrer Search API query parameter so usage analytics reports and dashboards are coherent.
     *
     * This value is optional, and will automatically try to resolve itself from the referrer search parameter.
     */
    originLevel3: string;
    /**
     * Optional analytics runtime environment, this is needed for analytics to work correctly if you're running outside of a browser.
     * See https://github.com/coveo/coveo.analytics.js for more info
     */
    runtimeEnvironment?: IRuntimeEnvironment;
  };
}

export const searchAPIEndpoint = '/rest/search/v2';
export const analyticsAPIEndpoint = '/rest/ua';

export const getConfigurationInitialState: () => ConfigurationState = () => ({
  organizationId: '',
  accessToken: '',
  platformUrl: platformUrl(),
  search: {
    apiBaseUrl: `${platformUrl()}${searchAPIEndpoint}`,
    locale: 'en-US',
  },
  analytics: {
    enabled: true,
    apiBaseUrl: `${platformUrl()}${analyticsAPIEndpoint}`,
    originLevel2: 'default',
    originLevel3: 'default',
  },
});
