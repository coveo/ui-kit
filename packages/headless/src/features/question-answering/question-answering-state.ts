import {QuestionAnswerDocumentIdentifier} from '../../api/search/search/question-answering';

export interface QuestionAnsweringRelatedQuestionState
  extends QuestionAnswerDocumentIdentifier {
  expanded: boolean;
}
export interface QuestionAnsweringState {
  /**
   * Determines if the snippet is liked, or upvoted by the end user.
   */
  liked: boolean;
  /**
   * Determines if the snippet is disliked, or downvoted by the end user.
   */
  disliked: boolean;
  /**
   * Determines if the snippet is expanded.
   */
  expanded: boolean;
  relatedQuestions: QuestionAnsweringRelatedQuestionState[];
}

export const getQuestionAnsweringInitialState: () => QuestionAnsweringState =
  () => ({
    liked: false,
    disliked: false,
    expanded: false,
    relatedQuestions: [],
  });
