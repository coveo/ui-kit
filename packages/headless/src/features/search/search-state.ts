import {SearchAPIErrorWithStatusCode} from '../../api/search/search-api-error-response';
import {Result} from '../../api/search/search/result';
import {SearchResponseSuccess} from '../../api/search/search/search-response';

export interface SearchState {
  /** The search response. For a full description, refer to {@link https://docs.coveo.com/en/13/cloud-v2-api-reference/search-api#operation/searchUsingPost}*/
  response: SearchResponseSuccess;
  /**
   * The time it took to execute the request against the Coveo platform.
   */
  duration: number;
  /**
   * The query (searchbox content) that was executed against the Coveo platform.
   */
  queryExecuted: string;
  /**
   * The error returned by the Coveo platform while executing the search request, if any. `null` otherwise.
   */
  error: SearchAPIErrorWithStatusCode | null;
  /**
   * Specifies if the query was automatically corrected by Headless.
   * This happens when there is no result returned by the API for a particular mispelling.
   */
  automaticallyCorrected: boolean;
  /**
   * `true` if the search request is currently being executed against the Coveo platform, `false` otherwise.
   */
  isLoading: boolean;
  /**
   * The list of results.
   */
  results: Result[];
  /**
   * The unique ID of the query, taken from the response's `searchUid. This value will not change when loading more results.
   */
  searchResponseId: string;
}

export function emptyQuestionAnswer() {
  return {
    answerSnippet: '',
    documentId: {
      contentIdKey: '',
      contentIdValue: '',
    },
    question: '',
    relatedQuestions: [],
    score: 0,
  };
}

export function getSearchInitialState(): SearchState {
  return {
    response: {
      results: [],
      searchUid: '',
      totalCountFiltered: 0,
      facets: [],
      queryCorrections: [],
      triggers: [],
      questionAnswer: emptyQuestionAnswer(),
    },
    duration: 0,
    queryExecuted: '',
    error: null,
    automaticallyCorrected: false,
    isLoading: false,
    results: [],
    searchResponseId: '',
  };
}
