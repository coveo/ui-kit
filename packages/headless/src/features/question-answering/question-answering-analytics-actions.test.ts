import {createRelay} from '@coveo/relay';
import {ThunkExtraArguments} from '../../app/thunk-extra-arguments';
import {
  buildMockSearchEngine,
  MockedSearchEngine,
} from '../../test/mock-engine-v2';
import {buildMockRaw} from '../../test/mock-raw';
import {buildMockResult} from '../../test/mock-result';
import {buildMockSearchResponse} from '../../test/mock-search-response';
import {buildMockSearchState} from '../../test/mock-search-state';
import {createMockState} from '../../test/mock-state';
import {getConfigurationInitialState} from '../configuration/configuration-state';
import {emptyQuestionAnswer} from '../search/search-state';
import {
  logExpandSmartSnippet,
  logExpandSmartSnippetSuggestion,
  logCollapseSmartSnippetSuggestion,
  logOpenSmartSnippetSuggestionSource,
  logOpenSmartSnippetSuggestionInlineLink,
  logCollapseSmartSnippet,
  logLikeSmartSnippet,
  logDislikeSmartSnippet,
  logOpenSmartSnippetSource,
  logOpenSmartSnippetInlineLink,
  logOpenSmartSnippetFeedbackModal,
  logCloseSmartSnippetFeedbackModal,
  logSmartSnippetFeedback,
  logSmartSnippetDetailedFeedback,
} from './question-answering-analytics-actions';
import {getQuestionAnsweringInitialState} from './question-answering-state';

const mockLogFunction = jest.fn();
const mockMakeExpandSmartSnippet = jest.fn(() => ({
  log: mockLogFunction,
}));
const mockMakeCollapseSmartSnippet = jest.fn(() => ({
  log: mockLogFunction,
}));
const mockMakeLikeSmartSnippet = jest.fn(() => ({
  log: mockLogFunction,
}));
const mockMakeDislikeSmartSnippet = jest.fn(() => ({
  log: mockLogFunction,
}));
const mockMakeOpenSmartSnippetSource = jest.fn((..._args) => ({
  log: mockLogFunction,
}));
const mockMakeOpenSmartSnippetInlineLink = jest.fn((..._args) => ({
  log: mockLogFunction,
}));
const mockMakeOpenSmartSnippetFeedbackModal = jest.fn(() => ({
  log: mockLogFunction,
}));
const mockMakeCloseSmartSnippetFeedbackModal = jest.fn(() => ({
  log: mockLogFunction,
}));
const mockMakeSmartSnippetFeedback = jest.fn((..._args) => ({
  log: mockLogFunction,
}));
const mockMakeSmartSnippetDetailedFeedback = jest.fn(() => ({
  log: mockLogFunction,
}));
const mockMakeExpandSmartSnippetSuggestion = jest.fn((..._args) => ({
  log: mockLogFunction,
}));
const mockMakeCollapseSmartSnippetSuggestion = jest.fn((..._args) => ({
  log: mockLogFunction,
}));
const mockMakeOpenSmartSnippetSuggestionSource = jest.fn((..._args) => ({
  log: mockLogFunction,
}));
const mockMakeOpenSmartSnippetSuggestionInlineLink = jest.fn((..._args) => ({
  log: mockLogFunction,
}));
const emit = jest.fn();

jest.mock('@coveo/relay');

jest.mocked(createRelay).mockReturnValue({
  emit,
  getMeta: jest.fn(),
  on: jest.fn(),
  off: jest.fn(),
  updateConfig: jest.fn(),
  clearStorage: jest.fn(),
  version: 'foo',
});

jest.mock('coveo.analytics', () => {
  const mockCoveoSearchPageClient = jest.fn(() => ({
    disable: jest.fn(),
    makeExpandSmartSnippet: mockMakeExpandSmartSnippet,
    makeCollapseSmartSnippet: mockMakeCollapseSmartSnippet,
    makeLikeSmartSnippet: mockMakeLikeSmartSnippet,
    makeDislikeSmartSnippet: mockMakeDislikeSmartSnippet,
    makeOpenSmartSnippetSource: mockMakeOpenSmartSnippetSource,
    makeOpenSmartSnippetInlineLink: mockMakeOpenSmartSnippetInlineLink,
    makeOpenSmartSnippetFeedbackModal: mockMakeOpenSmartSnippetFeedbackModal,
    makeCloseSmartSnippetFeedbackModal: mockMakeCloseSmartSnippetFeedbackModal,
    makeSmartSnippetFeedbackReason: mockMakeSmartSnippetFeedback,
    makeSmartSnippetDetailedFeedback: mockMakeSmartSnippetDetailedFeedback,
    makeExpandSmartSnippetSuggestion: mockMakeExpandSmartSnippetSuggestion,
    makeCollapseSmartSnippetSuggestion: mockMakeCollapseSmartSnippetSuggestion,
    makeOpenSmartSnippetSuggestionSource:
      mockMakeOpenSmartSnippetSuggestionSource,
    makeOpenSmartSnippetSuggestionInlineLink:
      mockMakeOpenSmartSnippetSuggestionInlineLink,
  }));

  return {
    CoveoSearchPageClient: mockCoveoSearchPageClient,
    history: {HistoryStore: jest.fn()},
  };
});

const exampleSearchUid = '456';
const exampleQuestion = 'Where am I?';
const exampleAnswerSnippet = 'You are here.';
const exampleScore = 9001;
const examplePermanentId = 'example permanent id';
const exampleQuestionAnswerId = 'foo';
const exampleFeedback = 'does_not_answer';
const exampleFeedbackDetails = 'example feedback details';

const exampleRelatedQuestion = {
  question: 'example question',
  answerSnippet: 'example answer',
  documentId: {
    contentIdKey: 'permanentid',
    contentIdValue: examplePermanentId,
  },
  score: 1,
};

const exampleQuestionAnswer = {
  question: exampleQuestion,
  answerSnippet: exampleAnswerSnippet,
  documentId: {
    contentIdKey: 'permanentid',
    contentIdValue: examplePermanentId,
  },
  score: exampleScore,
};

const exampleQuestionAnsweringState = {
  liked: false,
  disliked: false,
  expanded: false,
  feedbackModalOpen: false,
  questionAnswerId: 'bar',
};

const exampleRelatedQuestionState = {
  expanded: false,
  questionAnswerId: exampleQuestionAnswerId,
  contentIdKey: 'permanentid',
  contentIdValue: examplePermanentId,
};

const expectedDocumentInfo = {
  queryPipeline: '',
  documentUri: 'example documentUri',
  documentUriHash: 'example documentUriHash',
  collectionName: 'example collectionName',
  sourceName: 'example sourceName',
  documentPosition: 1,
  documentTitle: 'example documentTitle',
  documentUrl: 'example documentUrl',
  rankingModifier: 'example rankingModifier',
  documentAuthor: 'unknown',
};

const resultParams = {
  title: 'example documentTitle',
  uri: 'example documentUri',
  printableUri: 'printable-uri',
  clickUri: 'example documentUrl',
  rankingModifier: 'example rankingModifier',
  raw: buildMockRaw({
    urihash: 'example documentUriHash',
    source: 'example sourceName',
    collection: 'example collectionName',
    permanentid: examplePermanentId,
  }),
};
const exampleResult = buildMockResult(resultParams);

const exampleInlineLink = {
  linkText: 'example link text',
  linkURL: 'example link url',
};

describe('question answering analytics actions', () => {
  let engine: MockedSearchEngine;
  const searchState = buildMockSearchState({
    response: buildMockSearchResponse({
      searchUid: exampleSearchUid,
    }),
    results: [exampleResult],
    questionAnswer: {
      ...emptyQuestionAnswer(),
      ...exampleQuestionAnswer,
      relatedQuestions: [exampleRelatedQuestion],
    },
  });
  const questionAnsweringState = {
    ...getQuestionAnsweringInitialState(),
    ...exampleQuestionAnsweringState,
    relatedQuestions: [exampleRelatedQuestionState],
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('when analyticsMode is `legacy`', () => {
    beforeEach(() => {
      engine = buildMockSearchEngine(
        createMockState({
          configuration: {
            ...getConfigurationInitialState(),
            analytics: {
              ...getConfigurationInitialState().analytics,
              analyticsMode: 'legacy',
            },
          },
          search: searchState,
          questionAnswering: questionAnsweringState,
        })
      );
    });

    it('should log #logExpandSmartSnippet with the case payload', async () => {
      await logExpandSmartSnippet()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockMakeExpandSmartSnippet;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logCollapseSmartSnippet with the case payload', async () => {
      await logCollapseSmartSnippet()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockMakeCollapseSmartSnippet;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logLikeSmartSnippet with the case payload', async () => {
      await logLikeSmartSnippet()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockMakeLikeSmartSnippet;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logDislikeSmartSnippet with the case payload', async () => {
      await logDislikeSmartSnippet()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockMakeDislikeSmartSnippet;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logOpenSmartSnippetSource with the right payload', async () => {
      await logOpenSmartSnippetSource()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockMakeOpenSmartSnippetSource;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedDocumentInfo);
      expect(mockToUse.mock.calls[0][1]).toStrictEqual({
        contentIDKey: exampleQuestionAnswer.documentId.contentIdKey,
        contentIDValue: exampleQuestionAnswer.documentId.contentIdValue,
      });
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logOpenSmartSnippetInlineLink with the right payload', async () => {
      await logOpenSmartSnippetInlineLink(exampleInlineLink)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockMakeOpenSmartSnippetInlineLink;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedDocumentInfo);
      expect(mockToUse.mock.calls[0][1]).toStrictEqual({
        contentIDKey: exampleQuestionAnswer.documentId.contentIdKey,
        contentIDValue: exampleQuestionAnswer.documentId.contentIdValue,
        ...exampleInlineLink,
      });
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logOpenSmartSnippetFeedbackModal with the case payload', async () => {
      await logOpenSmartSnippetFeedbackModal()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockMakeOpenSmartSnippetFeedbackModal;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logCloseSmartSnippetFeedbackModal with the case payload', async () => {
      await logCloseSmartSnippetFeedbackModal()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockMakeCloseSmartSnippetFeedbackModal;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logSmartSnippetFeedback with the right payload', async () => {
      await logSmartSnippetFeedback(exampleFeedback)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const mockToUse = mockMakeSmartSnippetFeedback;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse.mock.calls[0][0]).toStrictEqual(exampleFeedback);
      expect(mockToUse.mock.calls[0][1]).toStrictEqual(undefined);
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logSmartSnippetDetailedFeedback with the right payload', async () => {
      await logSmartSnippetDetailedFeedback(exampleFeedbackDetails)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      const expectedFeedbackReason = 'other';

      const mockToUse = mockMakeSmartSnippetFeedback;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedFeedbackReason);
      expect(mockToUse.mock.calls[0][1]).toStrictEqual(exampleFeedbackDetails);
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logExpandSmartSnippetSuggestion with the right payload', async () => {
      await logExpandSmartSnippetSuggestion({
        questionAnswerId: exampleQuestionAnswerId,
      })()(engine.dispatch, () => engine.state, {} as ThunkExtraArguments);

      const expectedRelatedQuestionPayload = {
        question: exampleRelatedQuestion.question,
        answerSnippet: exampleRelatedQuestion.answerSnippet,
        documentId: exampleRelatedQuestion.documentId,
      };

      const mockToUse = mockMakeExpandSmartSnippetSuggestion;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse.mock.calls[0][0]).toStrictEqual(
        expectedRelatedQuestionPayload
      );
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logCollapseSmartSnippetSuggestion with the right payload', async () => {
      await logCollapseSmartSnippetSuggestion({
        questionAnswerId: exampleQuestionAnswerId,
      })()(engine.dispatch, () => engine.state, {} as ThunkExtraArguments);

      const expectedRelatedQuestionPayload = {
        question: exampleRelatedQuestion.question,
        answerSnippet: exampleRelatedQuestion.answerSnippet,
        documentId: exampleRelatedQuestion.documentId,
      };

      const mockToUse = mockMakeCollapseSmartSnippetSuggestion;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse.mock.calls[0][0]).toStrictEqual(
        expectedRelatedQuestionPayload
      );
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logOpenSmartSnippetSuggestionSource with the right payload', async () => {
      await logOpenSmartSnippetSuggestionSource({
        questionAnswerId: exampleQuestionAnswerId,
      })()(engine.dispatch, () => engine.state, {} as ThunkExtraArguments);

      const expectedRelatedQuestionPayload = {
        question: exampleRelatedQuestion.question,
        answerSnippet: exampleRelatedQuestion.answerSnippet,
        documentId: exampleRelatedQuestion.documentId,
      };

      const mockToUse = mockMakeOpenSmartSnippetSuggestionSource;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedDocumentInfo);
      expect(mockToUse.mock.calls[0][1]).toStrictEqual(
        expectedRelatedQuestionPayload
      );
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });

    it('should log #logOpenSmartSnippetSuggestionInlineLink with the right payload', async () => {
      await logOpenSmartSnippetSuggestionInlineLink(
        {questionAnswerId: exampleQuestionAnswerId},
        exampleInlineLink
      )()(engine.dispatch, () => engine.state, {} as ThunkExtraArguments);

      const expectedRelatedQuestionPayload = {
        question: exampleRelatedQuestion.question,
        answerSnippet: exampleRelatedQuestion.answerSnippet,
        documentId: exampleRelatedQuestion.documentId,
      };

      const mockToUse = mockMakeOpenSmartSnippetSuggestionInlineLink;
      expect(mockToUse).toHaveBeenCalledTimes(1);
      expect(mockToUse.mock.calls[0][0]).toStrictEqual(expectedDocumentInfo);
      expect(mockToUse.mock.calls[0][1]).toStrictEqual({
        ...expectedRelatedQuestionPayload,
        ...exampleInlineLink,
      });
      expect(mockLogFunction).toHaveBeenCalledTimes(1);
    });
  });

  describe('when analyticsMode is `next`', () => {
    beforeEach(() => {
      engine = buildMockSearchEngine(
        createMockState({
          configuration: {
            ...getConfigurationInitialState(),
            analytics: {
              ...getConfigurationInitialState().analytics,
              analyticsMode: 'next',
            },
          },
          search: searchState,
          questionAnswering: questionAnsweringState,
        })
      );
    });

    it('should log #logExpandSmartSnippet with the case payload', async () => {
      await logExpandSmartSnippet()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logCollapseSmartSnippet with the case payload', async () => {
      await logCollapseSmartSnippet()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logLikeSmartSnippet with the case payload', async () => {
      await logLikeSmartSnippet()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logDislikeSmartSnippet with the case payload', async () => {
      await logDislikeSmartSnippet()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logOpenSmartSnippetSource with the right payload', async () => {
      await logOpenSmartSnippetSource()()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logOpenSmartSnippetInlineLink with the right payload', async () => {
      await logOpenSmartSnippetInlineLink(exampleInlineLink)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logSmartSnippetFeedback with the right payload', async () => {
      await logSmartSnippetFeedback(exampleFeedback)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logSmartSnippetDetailedFeedback with the right payload', async () => {
      await logSmartSnippetDetailedFeedback(exampleFeedbackDetails)()(
        engine.dispatch,
        () => engine.state,
        {} as ThunkExtraArguments
      );

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logExpandSmartSnippetSuggestion with the right payload', async () => {
      await logExpandSmartSnippetSuggestion({
        questionAnswerId: exampleQuestionAnswerId,
      })()(engine.dispatch, () => engine.state, {} as ThunkExtraArguments);

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logCollapseSmartSnippetSuggestion with the right payload', async () => {
      await logCollapseSmartSnippetSuggestion({
        questionAnswerId: exampleQuestionAnswerId,
      })()(engine.dispatch, () => engine.state, {} as ThunkExtraArguments);

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logOpenSmartSnippetSuggestionSource with the right payload', async () => {
      await logOpenSmartSnippetSuggestionSource({
        questionAnswerId: exampleQuestionAnswerId,
      })()(engine.dispatch, () => engine.state, {} as ThunkExtraArguments);

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });

    it('should log #logOpenSmartSnippetSuggestionInlineLink with the right payload', async () => {
      await logOpenSmartSnippetSuggestionInlineLink(
        {questionAnswerId: exampleQuestionAnswerId},
        exampleInlineLink
      )()(engine.dispatch, () => engine.state, {} as ThunkExtraArguments);

      expect(emit).toHaveBeenCalledTimes(1);
      expect(emit.mock.calls[0]).toMatchSnapshot();
    });
  });
});
