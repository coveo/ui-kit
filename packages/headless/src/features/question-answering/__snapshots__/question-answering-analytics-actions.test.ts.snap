// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logCollapseSmartSnippet with the case payload 1`] = `
[
  "Qna.AnswerAction",
  {
    "action": "collapse",
    "answer": {
      "responseId": "456",
      "type": "SmartSnippet",
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logCollapseSmartSnippetSuggestion with the right payload 1`] = `
[
  "Qna.AnswerAction",
  {
    "action": "collapse",
    "answer": {
      "responseId": "456",
      "type": "SmartSnippetSuggestion",
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logDislikeSmartSnippet with the case payload 1`] = `
[
  "Qna.SubmitFeedback",
  {
    "answer": {
      "responseId": "456",
      "type": "SmartSnippet",
    },
    "feedback": {
      "liked": false,
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logExpandSmartSnippet with the case payload 1`] = `
[
  "Qna.AnswerAction",
  {
    "action": "expand",
    "answer": {
      "responseId": "456",
      "type": "SmartSnippet",
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logExpandSmartSnippetSuggestion with the right payload 1`] = `
[
  "Qna.AnswerAction",
  {
    "action": "expand",
    "answer": {
      "responseId": "456",
      "type": "SmartSnippetSuggestion",
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logLikeSmartSnippet with the case payload 1`] = `
[
  "Qna.SubmitFeedback",
  {
    "answer": {
      "responseId": "456",
      "type": "SmartSnippet",
    },
    "feedback": {
      "liked": true,
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logOpenSmartSnippetInlineLink with the right payload 1`] = `
[
  "Qna.CitationClick",
  {
    "answer": {
      "responseId": "456",
      "type": "SmartSnippet",
    },
    "citation": {
      "id": "example permanent id",
      "type": "InlineLink",
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logOpenSmartSnippetSource with the right payload 1`] = `
[
  "Qna.CitationClick",
  {
    "answer": {
      "responseId": "456",
      "type": "SmartSnippet",
    },
    "citation": {
      "id": "example permanent id",
      "type": "Source",
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logOpenSmartSnippetSuggestionInlineLink with the right payload 1`] = `
[
  "Qna.CitationClick",
  {
    "answer": {
      "responseId": "456",
      "type": "SmartSnippetSuggestion",
    },
    "citation": {
      "id": "example permanent id",
      "type": "InlineLink",
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logOpenSmartSnippetSuggestionSource with the right payload 1`] = `
[
  "Qna.CitationClick",
  {
    "answer": {
      "responseId": "456",
      "type": "SmartSnippetSuggestion",
    },
    "citation": {
      "id": "example permanent id",
      "type": "Source",
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logSmartSnippetDetailedFeedback with the right payload 1`] = `
[
  "Qna.SubmitFeedback",
  {
    "answer": {
      "responseId": "456",
      "type": "SmartSnippet",
    },
    "feedback": {
      "details": "example feedback details",
      "liked": false,
      "reason": "other",
    },
  },
]
`;

exports[`question answering analytics actions when analyticsMode is \`next\` should log #logSmartSnippetFeedback with the right payload 1`] = `
[
  "Qna.SubmitFeedback",
  {
    "answer": {
      "responseId": "456",
      "type": "SmartSnippet",
    },
    "feedback": {
      "liked": false,
      "reason": "does_not_answer",
    },
  },
]
`;
